### PR TITLE
Remove AutoDeleteSmallVector and AutoFreeSmallVector

### PR DESCRIPTION
**Pull request by michicc**
Remove the last remains of SmallVector. For the replacement, `std::unique_ptr` takes care of the memory management. Use of references and move semantics makes ownership clear and reduces potentials for memory leaks. Current compilers should be able to optimize most of the copy/moves out, resulting in very little runtime overhead.

This got a bit bigger than expected and picked up some unrelated fixes I saw when in the area.

Commits 41b6916, 0266dcb, and b27d0af could be split out if needed.

### CODE REVIEW

**Comment by michicc** [comment]
There's a slight work-around for clang-3.8 included (extra `std::move` that shouldn't be needed). Starting with clang-3.9 (verified with godbolt) this wouldn't be needed anymore.

**Comment by michicc** [comment]
Now with more `std::string` to please @nielsmh 😁 

Also, why the heck is GitHub displaying the commits in a mostly random order?

**Comment by M3Henry** [comment]
Oh good, this was on my TODO list

File: src/widgets/dropdown.cpp [hunk 1]
```diff
@@ -51,11 +51,11 @@ void DropDownListStringItem::Draw(int left, int right, int top, int bottom, bool
  * @return true if \a first precedes \a second.
  * @warning All items in the list need to be derivates of DropDownListStringItem.
  */
-/* static */ int DropDownListStringItem::NatSortFunc(const DropDownListItem * const *first, const DropDownListItem * const * second)
+/* static */ int DropDownListStringItem::NatSortFunc(std::unique_ptr<const DropDownListItem> const &first, std::unique_ptr<const DropDownListItem> const &second)
```
Comment (LordAro) [inline src/widgets/dropdown.cpp:54]: might make more sense to pass in the raw pointers here?

File: src/gfx_layout.cpp [hunk 2]
```diff
@@ -795,21 +808,21 @@ Point Layouter::GetCharPosition(const char *ch) const
  */
 const char *Layouter::GetCharAtPosition(int x) const
 {
-	const ParagraphLayouter::Line *line = this->front();
+	auto const &line = this->front();
```
Comment (LordAro) [inline src/gfx_layout.cpp]: const auto (+ others)

File: src/game/game_text.hpp [hunk 3]
```diff
@@ -29,12 +29,12 @@ struct LanguageStrings {
...
-	uint version;                  ///< The version of the language strings.
-	LanguageStrings *cur_language; ///< The current (compiled) language.
+	uint version;                                  ///< The version of the language strings.
+	std::shared_ptr<LanguageStrings> cur_language; ///< The current (compiled) language.
 
-	AutoDeleteSmallVector<LanguageStrings *> raw_strings;      ///< The raw strings per language, first must be English/the master language!.
-	AutoDeleteSmallVector<LanguageStrings *> compiled_strings; ///< The compiled strings per language, first must be English/the master language!.
-	StringList string_names;                                   ///< The names of the compiled strings.
+	std::vector<std::unique_ptr<LanguageStrings>> raw_strings;      ///< The raw strings per language, first must be English/the master language!.
+	std::vector<std::shared_ptr<LanguageStrings>> compiled_strings; ///< The compiled strings per language, first must be English/the master language!.
```
Comment (LordAro) [inline src/game/game_text.hpp:36]: how difficult would it be to "fix" the use of shared_ptr here?
or, perhaps the various places that "own" the pointer could be documented?

**Review by LordAro — COMMENTED** [review]
use of noexcept is interesting. I've seen varying opinions on how worthwhile it actually is...

File: src/widgets/dropdown.cpp (hunk 1 above)
Comment (michicc) [reply src/widgets/dropdown.cpp:54]: It's not about making sense, it's about what `std::sort` expects. And in this usage, a raw pointer won't work.

File: src/game/game_text.hpp (hunk 3 above)
Comment (michicc) [reply src/game/game_text.hpp:36]: Very easily, it is only `GameStrings::cur_language` that points to one of the compiled string entries. We store a raw pointer now and just assume it's never going to be invalidated; still possible.

In pure though, raw pointers are bad and should never be used. Of course, the codebase is far, far, far from that point now. So, philosophical question time: do the "right" thing, or the pragmatic thing?

File: src/gfx_layout.cpp (hunk 2 above)
Comment (michicc) [reply src/gfx_layout.cpp]: Will do.

**Comment by michicc** [comment]
Rebased to drop the already applied OSX commit.

**Comment by PeterN** [comment]
Does the change from returning a `DropDownList *` to returning a `DropDownList` mean it makes a copy?

**Comment by michicc** [comment]
In the literal sense: No. `std::vector` is a moveable type, which means the return statement will resolve to a move assignment and not a copy assignment. Moving a vector is quite cheap.

In the less literal sense: Many current compilers will perform copy elision where possible, which means even the move assignment is likely to be optimized out.

**Review by LordAro — APPROVED** [review]

**Pull request merged**